### ISSUE TITLE
Wintersmith as a Frontend Development Tool

### ISSUE THREAD

**Issue by airtonix** [jnordberg/wintersmith#270]
tl;dr : create another plugin type that gets run in the requestHandler of the preview server only.

Where I work, I am being charged with creating a frontend development patternlab framework for all our projects.

We have a process that goes : `discovery >  define > design > develop[frontend] > develop[backend] > deploy`, our patternlab would cover `define > design > develop` stages.

My experience in full stack development from the world of Django has taught me that reducing duplication of workload increases quality and reduces maintenance. To this end it would become the canonical source of backend templates created and approved during `design > develop[frontend]`.

The patternlab I have created is based on wintersmith, uses liquid templates and follows the Atomic Design principles.

But we require api mocking (as in lineman), but we also require the contentTree that wintersmith provides.

I tried to provide a restful mocking via existing means:
- generators: can't because i'd have to generate all possible permutations to be mocked.
- contentplugins: can't because url routes would only be able to handle the list paradigm `/api/users/` and not dynamic routes like `/api/users/:id`
- views: essentially the same limitations as contentplugins.

I propose the creation of another plugin type: `MiddlewarePlugin`
- We load them like other plugins `config.plugins = [ ... './src/plugins/RestfulApi' ... ]`
- They get registered into `env.middlewarePlugins`
- They're only used with the `wintersmith preview` feature
- They're instantiated outside of the `requestHandler`
- They expose a `.dispatch` method
- The `dispatch` method is called in the `requestHandler` before `contentHandler`

``` coffee

class RestfulMiddlewarePlugin extends env.MiddlewarePlugin
    constructor: ->
        @resources = []
    dispatch: (request, response, next) ->
        response.end JSON.stringify data: @resources
```
